cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -595,7 +598,13 @@ struct GrouperFastImpl : public Grouper {
     return std::move(impl);
   }
 
-  ~GrouperFastImpl() { map_.cleanup(); }
+  Status Reset() override {
+    rows_.Clean();
+    rows_minibatch_.Clean();
+    map_.cleanup();
+    RETURN_NOT_OK(map_.init(encode_ctx_.hardware_flags, ctx_->memory_pool()));
+    return Status::OK();
```

Comment:
Do you think we should enlarge the size of the `temp_stack_`? Or the stack's internal states like `top_` may no enough for the next bunch of operations after grouper reset.

Reply:
I think all the temp vectors allocated by the `temp_stack_` would be released at the end the of the call to each individual grouper member function (IMO that's why they are designed to be "temp vector"). Meaning by the time of `Reset`, the stack is empty already.

Reply:
Ah, i see, thanks.
`TempVectorStack` will hold a resizable buffer allocated in `TempVectorStack::Init` function until the end of the grouper object.
But the `TempVectorHolder` will release the used space (already allocated in TempVectorStack ) at the call to each of individual grouper member function as you said.

Reply:
Can you add a DCHECK that the `temp_stack_` is empty, then?

Reply:
Yeah that's probably reasonable. But considering current temp stack doesn't have a public method for size/empty check, and I have other open PRs for temp stack restructure, I'd add the necessary methods after other PRs are done and get this one rebased.